atom/common/promise_util.h

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2018 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#ifndef ATOM_COMMON_PROMISE_UTIL_H_
+#define ATOM_COMMON_PROMISE_UTIL_H_
+
+#include <string>
+
+#include "native_mate/converter.h"
+
+namespace atom {
+
+namespace util {
+
+class Promise {
+ public:
+  explicit Promise(v8::Isolate* isolate) {
+    isolate_ = isolate;
+    resolver_.Reset(isolate, v8::Promise::Resolver::New(isolate));
+  }
+  ~Promise() { resolver_.Reset(); }
+
+  v8::Isolate* isolate() const { return isolate_; }
+
+  virtual v8::Local<v8::Object> GetHandle() const;
+
+  Promise* Resolve() {
+    GetInner()->Resolve(v8::Undefined(isolate()));
+    return this;
```

Comment:
Why does this (and its siblings) return a `Promise*` instead of `void` or `Maybe<bool>`?

Reply:
Tbh was trying to be similar to the JS Promise apis that return promises from similar methods (Promise.resolve) Etc.

Makes it easy to return a resolved promise.  Happy to reconsider if objectionable :smile:

Reply:
I see that argument and agree with you that it's useful, but yeah I feel uncomfortable returning an unmanaged pointer to `this`. This idiom works better in JS because there are no unmanaged pointers there :smiley:

Reply:
Discussed this IRL.  Basically determined that the lifetime of a promise should be managed by the thing receiving it not the thing creating it.  I.e. Whatever receives this returned value needs to manage it's lifetime.  Either by returning it to JS land and letting the user manage the lifetime or by managing in cpp land